rest_api/controller/documents.py

```diff
@@ -0,0 +1,107 @@
+from datetime import datetime
+from haystack.schema import Document
+from haystack.document_store import sql
+from rest_api.config import LOG_LEVEL
+from typing import Dict
+from fastapi import APIRouter, HTTPException
+import logging
+
+sql = sql.SQLDocumentStore()
+now = datetime.now()
+SUCCESS: int = 200
+CREATED: int = 201
+
+logger = logging.getLogger('haystack')
+logger.setLevel(LOG_LEVEL)
+
+router = APIRouter()
+
+
+@router.post("/documents", status_code=CREATED)
+def create_document(payload: Dict[Document]):
+    """
+    Creates a Document object
+    :param payload: input from the client to create a document.
+    """
+    logger.info("Constructing document object")
+    created_document = Document.from_dict(payload)
+    logger.info("Constructed!")
+    return {"created": created_document,
+            "created_at": now}
+
+
+@router.get("/documents", status_code=SUCCESS)
+def get_all_documents(start: int = 0, limit: int = 20):
+    """
+    Returns the documents which are available in the database
+    Supports pagination
+    :param start: used to set the starting point
+    :param limit: used to delimit the number of retrieved documents
+    """
+    try:
+        logger.info("Retrieving documents")
+        retrieved_documents = sql.get_all_documents()
+        return retrieved_documents[start: start + limit]
+    except Exception as e:
+        if TimeoutError:
+            raise Exception("Request has timed out")
+        else:
+            raise Exception(f"Something went wrong while retrieving documents", e.message)
+
+
+@router.get("/documents/{id}", status_code=SUCCESS)
+def get_document(id: str):
+    """
+    Returns a document given its id
+    :param id: UUID pointing to the document to retrieve
+    """
+    try:
+        retrieved_document = sql.get_document_by_id(id)
+        if not retrieved_document:
+            raise HTTPException(status_code=404, detail=f"No document with id {id} could be found")
```

Comment:
Exception raise here will be caught by line no 63, and that will convert it into normal Exception at line no 67

Reply:
HTTPException is not a generic Python exception, meaning an exception will be raised and the code will terminate with exit code != 0. We won't go further in the calling chain

Reply:
Not sure about but FastAPI HTTPException is sub class of Exception.
https://github.com/tiangolo/fastapi/blob/4d208b2b9035e24bdf80505571b5b1bac8f9ae7a/fastapi/exceptions.py#L8
https://github.com/encode/starlette/blob/bfa61ad92b63b6929e485f2189fd299f94189b09/starlette/exceptions.py